subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -155,57 +151,52 @@ private CompleteFileSystemLocationSnapshot snapshot(String location) {
-                updateRoot((root, changeListener) -> root.store(regularFileSnapshot.getAbsolutePath(), regularFileSnapshot, changeListener));
+                virtualFileSystem.update((root, changeListener) -> root.store(regularFileSnapshot.getAbsolutePath(), regularFileSnapshot, changeListener));
                 return regularFileSnapshot;
             case Missing:
                 MissingFileSnapshot missingFileSnapshot = new MissingFileSnapshot(location, fileMetadata.getAccessType());
-                updateRoot((root, changeListener) -> root.store(missingFileSnapshot.getAbsolutePath(), missingFileSnapshot, changeListener));
+                virtualFileSystem.update((root, changeListener) -> root.store(missingFileSnapshot.getAbsolutePath(), missingFileSnapshot, changeListener));
                 return missingFileSnapshot;
             case Directory:
                 CompleteFileSystemLocationSnapshot directorySnapshot = directorySnapshotter.snapshot(location, null, new AtomicBoolean(false));
-                updateRoot((root, changeListener) -> root.store(directorySnapshot.getAbsolutePath(), directorySnapshot, changeListener));
+                virtualFileSystem.update((root, changeListener) -> root.store(directorySnapshot.getAbsolutePath(), directorySnapshot, changeListener));
                 return directorySnapshot;
             default:
                 throw new UnsupportedOperationException();
         }
     }
 
-    private void updateRoot(SnapshotHierarchy.DiffCapturingUpdateFunction updateFunction) {
-        root.update(updateFunctionDecorator.decorate(updateFunction));
-    }
-
     private CompleteFileSystemLocationSnapshot readLocation(String location) {
-        return root.get().getSnapshot(location)
+        return virtualFileSystem.get().getSnapshot(location)
             .orElseGet(() -> producingSnapshots.guardByKey(location,
-                () -> root.get().getSnapshot(location).orElseGet(() -> snapshot(location)))
+                () -> virtualFileSystem.get().getSnapshot(location).orElseGet(() -> snapshot(location)))
             );
     }
 
     @Override
-    public void update(Iterable<String> locations, Runnable action) {
-        updateListener.locationsUpdated(locations);
-        root.update(root -> {
-            SnapshotHierarchy result = root;
+    public void write(Iterable<String> locations, Runnable action) {
+        writeListener.locationsWritten(locations);
+        virtualFileSystem.update((outerRoot, outerDiffListener) -> {
             for (String location : locations) {
-                result = updateFunctionDecorator.decorate((currentRoot, changeListener) -> currentRoot.invalidate(location, changeListener)).updateRoot(result);
+                virtualFileSystem.update((innerRoot, diffListener) -> innerRoot.invalidate(location, diffListener));
             }
-            return result;
+            return virtualFileSystem.get();
         });
```

Comment:
We wanted to do the whole invalidate operation atomically. Currently, it is also necessary to only have one call on the root per `update()` call, since that is all the listener can handle. This was part of the reason why I moved to having a version of the root which doesn't pass in the listener in the other PR.